### PR TITLE
206 make all instructions immutable

### PR DESCRIPTION
**Pull request by flamion**

### CODE REVIEW

**Review by PietHelzel — CHANGES_REQUESTED** [review]

File: backend/shork/src/main/kotlin/internal/instruction/AbstractInstruction.kt [hunk 1]
```diff
@@ -31,6 +44,84 @@ internal abstract class AbstractInstruction(
...
 
+    /**
+     * Creates a new instance of the instruction with the changes made by the provided block.
+     * Returns the new instruction and whatever the block returns. Does not write the changes to
+     * memory.
+     *
+     * @param block The (lambda) function that makes the changes to the fields
+     * @return A pair of the new instruction and whatever the block returns
+     */
+    fun <T> write(block: (MutableAbstractInstruction) -> T): Pair<AbstractInstruction, T> {
```
[the remark this answers is not in the document]
Comment (flamion) [reply backend/shork/src/main/kotlin/internal/instruction/AbstractInstruction.kt:55]: How about `writeToMemory`?

File: backend/shork/src/main/kotlin/internal/instruction/AbstractInstruction.kt [hunk 2]
```diff
@@ -3,15 +3,28 @@ package software.shonk.interpreter.internal.instruction
...
+ * Mutable version of [AbstractInstruction]. Used to collect changes for copying. Exists only to
+ * make copying instruction changes more convenient
+ */
+internal data class MutableAbstractInstruction(
     var aField: Int,
     var bField: Int,
     var addressModeA: AddressMode,
     var addressModeB: AddressMode,
     var modifier: Modifier,
+)
```
[the remark this answers is not in the document]
Comment (flamion) [reply backend/shork/src/main/kotlin/internal/instruction/AbstractInstruction.kt:19]: Yeah, I had the same thought but couldn't come up with a better name

File: backend/shork/src/main/kotlin/internal/instruction/AbstractInstruction.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (PietHelzel) [reply backend/shork/src/main/kotlin/internal/instruction/AbstractInstruction.kt:55]: Sounds good

**Review by PietHelzel — APPROVED** [review]
lgtm

**Pull request merged**